pkg/helm/helm.go

```diff
@@ -18,165 +18,163 @@ package helm
+	// configuration.
+	chartCfg := repo.Entry{
+		Name: "stable",
+		URL:  repoURL,
+	}
+	if chartCfg.URL == "" {
+		chartCfg.URL = stableURL
+	}
+	chartRepo, err := c.repositoryClient.Initialize(&chartCfg, getter.All(c.settings))
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	klog.V(3).Infof("helm client: caching stable repository")
-	cif := c.home.CacheIndex("stable")
-	cr := repo.Entry{
-		Name:  "stable",
-		Cache: cif,
-		URL:   c.repoURL,
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: downloading index file")
 	}
-
-	var settings environment.EnvSettings
-	r, err := repo.NewChartRepository(&cr, getter.All(settings))
+	indexPath, err := c.repositoryClient.DownloadIndex(chartRepo)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	if err := r.DownloadIndexFile(c.home.Cache()); err != nil {
-		return errors.Wrapf(err, "Looks like %q is not a valid chart repository or cannot be reached", cr.URL)
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: loading repository")
+	}
+	indexFile, err := c.repositoryClient.Load(indexPath)
+	if err != nil {
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	f.Update(&cr)
-	f.WriteFile(c.home.RepositoryFile(), 0644)
-
-	// Load the repositories.yaml
-	c.rf, err = repo.LoadRepositoriesFile(c.home.RepositoryFile())
+	chartRepo.IndexFile = indexFile
+	c.SetChartRepo(chartRepo)
 
-	klog.V(5).Infof("helm client: initialized client")
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: successfully initialized")
+	}
 
-	return err
+	return nil
 }
 
-func (c *Client) ListCharts() (map[string]repo.ChartVersions, error) {
-	klog.V(4).Infof("helm client: listing charts")
-
-	charts := map[string]repo.ChartVersions{}
-
-	// TODO: handle non-unique names across repos
-	for _, r := range c.rf.Repositories {
-		n := r.Name
-		f := c.home.CacheIndex(n)
-		index, err := repo.LoadIndexFile(f)
-		if err != nil {
-			return nil, errors.Wrapf(err, "Could not load helm repository index at %s", f)
-		}
+// SetChartRepo allows setting the chart repository being used by the client.
+func (c *Client) SetChartRepo(chartRepo *repo.ChartRepository) {
+	c.chartRepo = chartRepo
+}
 
-		for chart, chartVersions := range index.Entries {
-			charts[chart] = chartVersions
-		}
+// ListCharts lists the charts from the chart repository.
+func (c *Client) ListCharts() map[string]repo.ChartVersions {
+	if l := c.log.V(4).Get(); l != nil {
+		l.Log("helm client: listing charts from %s", c.chartRepo.Config.URL)
+		defer l.Log("helm client: listed charts from %s", c.chartRepo.Config.URL)
 	}
-
-	klog.V(4).Infof("helm client: listed charts")
-
-	return charts, nil
+	return c.chartRepo.IndexFile.Entries
 }
 
+// GetChart gets a chart that exists in the chart repository.
 func (c *Client) GetChart(name, version string) (*repo.ChartVersion, error) {
```

Comment:
I just realized that we're (and have been before the PR) matching on _app_ version, not _chart_ version; [`IndexFile.Get()`](https://pkg.go.dev/helm.sh/helm/v3@v3.2.3/pkg/repo?tab=doc#IndexFile.Get) work on _chart_ versions, and that's an important reason we can't use it.

We should probably document this — both in the method doc comment, and then all the way up its callers…

Yes, that documentation is important enough that I'd prefer to do it in this PR, even though technically it's an existing issue :(

Reply:
Fixed with https://github.com/kubernetes-sigs/minibroker/pull/72/commits/d63a359db356f322a336d438139f32ee8dc9ed05.